src/main/java/com/mraof/minestuck/entity/AnimatedCreatureEntity.java

```diff
@@ -0,0 +1,165 @@
+package com.mraof.minestuck.entity;
+
+import net.minecraft.entity.CreatureEntity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.network.datasync.DataParameter;
+import net.minecraft.network.datasync.DataSerializers;
+import net.minecraft.network.datasync.EntityDataManager;
+import net.minecraft.world.World;
+import software.bernie.geckolib3.core.IAnimatable;
+import software.bernie.geckolib3.core.controller.AnimationController;
+import software.bernie.geckolib3.core.manager.AnimationFactory;
+
+public abstract class AnimatedCreatureEntity extends CreatureEntity implements IAnimatable {
+    private final AnimationFactory factory = new AnimationFactory(this);
+    protected static final DataParameter<Integer> CURRENT_ACTION = EntityDataManager.defineId(AnimatedCreatureEntity.class, DataSerializers.INT);
+
+    private int heavyAttackTicks = 0;
+    private int recoveryTicks = 0;
+    private int attackDelay;
+    private int attackRecovery;
+    private int actionResetTicks;
+
+    protected boolean canMoveWhileAttacking = false;
```

Comment:
Not sure if this is necessary, but if you want then add a comment for this line explaining that if this variable is set to true, animators need to ensure that the bones used in the attack animation needs to not interfere with the bones used in the walk animation. Riotmode has separated walking into an animation called walk and another called walk arms, the second one used to animate entity's arms so that the base walk animation would not interfere with other animations in cases like these

Reply:
I see what you mean but im not sure this is the place to put that kind of information. it should be visible to the whole team and not buried in code. if you have a look at the imp entity, you need to code in the walk/attack/arms behavior anyways

Reply:
reasonable